codice/ddf
# DDF-2251 Refactor Registry Report Builder to be able to generate 'summaries' of parts of the overall report

#### What does this PR do?

In general, the registry report now utilizes the maps generated from the `registry-schema-bindings` and through `RegistryReportBuilder`, the values in the maps populate a string of html that is returned and compiled through`RegistryRestEndpoint`.  

The new reports can be accessed through the endpoints: `.../services/registries/{registryID}/report` , `.../report/summary` , and `.../report/organizations` . Further description is presented in background context. 

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?

@gordocanchola @clockard @mcalcote @vinamartin @brianfelix 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).

@pklinef @shaundmorris @figliold

#### How should this be tested?

Start DDF Registry 
Go to DDF Catalog 
Click the Identity Node - Must add a service, add attributes to the identity node 
Go to sources tab , click corresponding identity node 
Click Registry Source Report 
Ensure that all fields you entered have been filled 
For Summary and Organizations Report, insert the corresponding URL (/report/summary and /report/organizations) and add Organization attributes 

#### Any background context you want to provide?

There are three different types of html strings that can be generated which are determined based on the endpoints—two new endpoints (summary and section) were added in `RegistryRestEndpoint`. Descriptions of each endpoint are below: 

1.  `/report`: Utilizes the `report.hbt` handlebar template where all entities (Node Info, Services, Organizations, Contacts, and Collections) of the registry node are displayed based on the map values generated from the class in the schema-bindings `RegistryObjectListConverter`

2. `/report/summary`:  Utilizes the `summary.hbt` handlebar template where one the entities of an organization (Point of Contact, Name, Address, Email, Phone Number) is displayed based on the map values extracted from the metacard. 

3. `/report/{section}`: {section} gives freedom to extract certain sections of the report, however specific handlebar templates and additional methods must be made to accommodate for this. Currently Organizations has been extracted from the report.  

- `/report/organizations`: Utilizes the `organizations.hbt` handlebar template where all of the organizations and their entities (ids, Addresses, Emails, Phone Numbers. etc. ) are  displayed based on the map values generated from the class in the schema-bindings `OrganizationWebConverter`

#### What are the relevant tickets?

[DDF-2251](https://codice.atlassian.net/browse/DDF-2251)
[DDF-2336 ](https://codice.atlassian.net/browse/DDF-2336)(blocked by this ticket, will provide an action provider to allow the summary report to be shown directly on the UI)
[DDF-2234](https://codice.atlassian.net/browse/DDF-2334) (Shows null values in summary report when certain optional fields are not filled in) 

#### Screenshots (if appropriate)

**Full Report** 

![screen shot 2016-07-26 at 1 56 46 pm](https://cloud.githubusercontent.com/assets/5401992/17155681/02941f1a-533b-11e6-832b-859ba29dffe9.png)

**Summary Report** 

![screen shot 2016-07-26 at 1 56 29 pm](https://cloud.githubusercontent.com/assets/5401992/17155647/dbb39c04-533a-11e6-87d0-20f0880c3452.png)

When null values are present, should be fixed with DDF-2235: 

![screen shot 2016-07-26 at 2 17 21 pm](https://cloud.githubusercontent.com/assets/5401992/17155851/b3fb6632-533b-11e6-9e60-77548d6d7ff8.png)

**Organizations Report** 

![screen shot 2016-07-26 at 1 56 15 pm](https://cloud.githubusercontent.com/assets/5401992/17155676/fc41f63c-533a-11e6-8fed-c75b99ad6cfd.png)

#### Checklist:
